CRM/Commitcivi/Logic/StripeMigration.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+class CRM_Commitcivi_Logic_StripeMigration {
+    
+
+    /**
+     * @param \CRM_Commitcivi_Model_Event $event
+     *
+     * @return integer
+     * @throws \CiviCRM_API3_Exception
+     */
+    public function migrate(CRM_Commitcivi_Model_Event $event)
+    {
+        $contact = $event->contact;
+        $donation = $event->donation;
+        
+        # find the current houdini recurring donation using email, processor, amount and start_date
+        $query = ""
+            . "SELECT recur.id recurring_id, contact.id contact_id "
+            . " FROM civicrm_contribution_recur recur "
+            . " JOIN civicrm_contact contact ON (contact.id=recur.contact_id) "
+            . " JOIN civicrm_email email ON (contact.id=email.contact_id AND email.is_primary)"
+            . "WHERE email.email = %1 "
+            . "AND payment_processor_id = 11 "
+            . "AND amount * 100 = %2 "  # units are 1s in Civi (3 euros) and 100ths (300) in Houdini 
+            . "AND left(start_date, 10) = %3 ";
+
+        $query_params = [
+            '1' => [$contact->email, 'String'],
+            '2' => [$donation->amount, 'Float'],
+            '3' => [$donation->stripeRecurringStart, 'String']
+        ];
+
+        $result = CRM_Core_DAO::executeQuery($query, $query_params);
+        $migrated = 0;
+
+        while ($result->fetch()) {
+            civicrm_api3('StripeSubscription', 'import', [
+                'subscription_id' => $donation->stripeSubscriptionId,
+                'contact_id' => $result->contact_id,
+                'payment_processor_id' => 1  # NOTE: Verify!
+            ]);
+            civicrm_api3(
+                'ContributionRecur',
+                'cancel',
+                [
+                    'id' => $result->recurring_id,
+                    'cancel_reason' => "Migrated to Stripe {$donation->stripeSubscriptionId}"
+                ]
+            );
+
+            ++$migrated;
+        }
+
+        //
+        // How do we report a warning without exiting the consumer?
+        //
+        return $migrated == 0 ? "Didn't find any recurring donations to migrate." : 1;
```

Comment:
I'm not sure what to do here - this return code seems to cause the consumer process to exit. Reading the code - I don't think it should.

We need to have a log of subscriptions that were imported where we didn't find a recurring donation to cancel. But it's not that big a deal, since CiviCRM doesn't charge folks. 

How do log a warning / info and put these in a error queue without exiting?

Reply:
This behaviour is because historically, we rarely had errors caused by the content of the message, rather because of a network / infrastructure problem that a reboot of the consumer would often solve. So the consumers exits so that the supervisor restarts it.
It's true that this assumption is not so true anymore, and more intelligence could be added to the error handling. But in the meantime, I guess this can't be escaped and you need to manually restart or use a supervisor.

Reply:
OK - will leave it as is then.